src/compiler/program.ts

```diff
@@ -2774,8 +2802,12 @@ namespace ts {
             refFile?: RefFile
         ): void {
             tracing.push(tracing.Phase.Program, "processTypeReferenceDirective", { directive: typeReferenceDirective, hasResolved: !!resolveModuleNamesReusingOldState, refKind: refFile?.kind, refPath: refFile?.file.path });
-            processTypeReferenceDirectiveWorker(typeReferenceDirective, resolvedTypeReferenceDirective, refFile);
-            tracing.pop();
+            try {
+                processTypeReferenceDirectiveWorker(typeReferenceDirective, resolvedTypeReferenceDirective, refFile);
+            }
+            finally {

```

Comment:
I am concerned about amount of try catch blocks we have added as part of this. 
On tsserver we already do executeCommand under try catch block. I think it would be better if we didnt have to add try catch everywhere but to ensure tracing stack size in the finally block of onMessage by popping till given index at the time of push

Reply:
Averaged over a million runs, `F3` is the same speed as `F1` and `F4` is the same speed as `F2`:

```js
function F1(array, index, increment) {
    array[index] += increment;
    array[index] += increment;
}

function F2(array, index, increment) {
    array[index] += increment;
    array[index] += increment;
    throw new Error("F2");
}

function F3(array, index, increment) {
    try {
        array[index] += increment;
    }
    finally {
        array[index] += increment;
    }
}

function F4(array, index, increment) {
    try {
        array[index] += increment;
        throw new Error("F3");
    }
    finally {
        array[index] += increment;
    }
}
```

Having said that, `popAll` is kind of cute, so I'm open to that approach if you feel strongly about it.

Reply:
i prefer popAlll as its easier in future as well to not foreget try catch additions.

Reply:
So, I added an assert that the stack is empty at the beginning of `onMessage` and it failed.  My guess is that the remaining entries are from a cancelled diagnostic request (which happens between requests in VS Code, right?).

I also had to leave the try-finally for each begin-end pair, because they don't use a stack.

Given that neither the lab perf test nor my local microbenchmark showed a regression from the try-finally requests, I'm inclined to leave the try-finally blocks and add a comment to `pop` saying that it should be in a finally block.

Thoughts?

Reply:
I posted an update with it seemingly working after putting a try-finally around the background work code.  It seems pretty fragile to me, but it does eliminate nearly all new try-finally blocks from tsc.

Reply:
The change now introduces no new try-finally blocks.